### PR TITLE
Add generic system interface for ros2_control

### PR DESCRIPTION
**Pull request by destogl**
This is a overview PR with integration of ros2_canopen and ros2_control. The PR will be divided into smaller PR to make review easier. Concrete PR will be referenced below.

- Updates of the `DeviceManager` to support its use as library
- Updates with automated tests of `DeviceManager`
- Addition of `canopen_ros2_control::CanopenSystem`
- Addition of  `canopen_ros2_controllers::CanopenController`
- Addition of `canopen_ros2_control::CIA402System`
- Usage examples for `CIA402System`

### CODE REVIEW

**Comment by ipa-cmh** [comment]
Regarding the state machine matching I would propose the following:
image

**Activating** should call *handleInit()*. This will put the motion controller into Operation Enabled state and executed the configured homing sequence. Then it should call *enterModeAndWait(MotorBase::??)* with ?? either No_Mode or the Mode user configured during Inactive.

**Deactivating** should set call *enterModeAndWait(MotorBase::No_Mode)* and then *handleShutdown()*.

In this, we do not use QuickStop which is usually used for stopping the device during operation fast. This does only Stop the movement and usually slam in the breaks. It does not disable torque. I am not sure, how this would be integrated into the ros2_control interface.

**Comment by ipa-cmh** [comment]
There is some major restructuring being done in #34.
After some analysis, it seems that it is easier to rebase this PR onto #34. 
It is mainly a renaming and CMakeLists.txt merge issue.

This is the result of the rebase, if possible please check: https://github.com/ipa-cmh/ros2_canopen/tree/canopen-system-interface

I am planning to do tidy up the device interface as well, but I will wait until this PR is merged.

**Review by ipa-cmh — CHANGES_REQUESTED** [review]
I did an intial review. Generally it looks good, I suppose we just need to remove the tests from CMakeLists.txt for now.

File: .gitignore
Comment (destogl) [inline .gitignore:6]:
```suggestion
**/master.dcf
```

**Comment by destogl** [comment]
@livanov93 do we need all of those `*.dcf` files?

File: canopen_core/CMakeLists.txt
Comment (ipa-cmh) [inline canopen_core/CMakeLists.txt:255]: This should proably be device_container.

**Review by ipa-cmh — CHANGES_REQUESTED** [review]

**Comment by ipa-cmh** [comment]
This PR breaks compatability with foxy and galactic, as far as I can see. I suppose there is no way of having them supported easily.
For me this means, we I need to split of foxy and galactic support into a seperate branch.

**Comment by livanov93** [comment]
> This PR breaks compatability with foxy and galactic, as far as I can see. I suppose there is no way of having them supported easily. For me this means, we I need to split of foxy and galactic support into a seperate branch.

We will need split branches for separate `ROS_DISTRO`s because the same approach is used in [`ros2_control`](https://github.com/ros-controls/ros2_control/branches).

**Comment by borongyuan** [comment]
> > This PR breaks compatability with foxy and galactic, as far as I can see. I suppose there is no way of having them supported easily. For me this means, we I need to split of foxy and galactic support into a seperate branch.
> 
> We will need split branches for separate `ROS_DISTRO`s because the same approach is used in [`ros2_control`](https://github.com/ros-controls/ros2_control/branches).

This is reasonable. How to migrate Hardware Components from Foxy to Galactic is explained [here](https://control.ros.org/master/doc/ros2_control/hardware_interface/doc/hardware_components_userdoc.html#migration-from-foxy-to-galactic).

**Comment by ipa-cmh** [comment]
Alright. I have create seperate branches for foxy, galactic, humble. I'd host rolling in master. I have reserved some lab time tomorrow morning for testing this with a the trinamic pd42 device and maybe the prbt.

**Comment by ipa-cmh** [comment]
@livanov93 @destogl I would be okay with only supporting ros2_control in humble onwards. Foxy is EOL in about 8 month and Galactic in November anyways.

**Comment by destogl** [comment]
> nly supporting ros2_control in humble onwards. Foxy is EOL in about 8 month and Galactic in November anyways.

yes of course! This saves us some time for sure.

File: canopen_core/CMakeLists.txt
Comment (destogl) [inline canopen_core/CMakeLists.txt:255]:
```suggestion
  device_container
```

**Review by ipa-cmh — APPROVED** [review]

**Comment by ipa-cmh** [comment]
@destogl @livanov93
Finished tests and did some updates to integrate better with the current launch infrastructure.
You can find a few changes in https://github.com/StoglRobotics-forks/ros2_canopen/pull/6

**Comment by livanov93** [comment]
@ipa-cmh I did part of that in the ros-industrial/ros2_canopen@02c7090. Can you rebase in StoglRobotics-forks/ros2_canopen#6?

**Comment by livanov93** [comment]
> @destogl @livanov93 Finished tests and did some updates to integrate better with the current launch infrastructure. You can find a few changes in [StoglRobotics-forks#6](https://github.com/StoglRobotics-forks/ros2_canopen/pull/6)

@ipa-cmh I did part of that in the ros-industrial/ros2_canopen@02c7090. Can you rebase in StoglRobotics-forks/ros2_canopen#6?

**Comment by ipa-cmh** [comment]
@livanov93 done. basically this moves testing specific device files into the canopen_tests folder.

**Comment by destogl** [comment]
@ipa-cmh @livanov93 I have added a short documentation about testing this interface.

To not block this PR any longer I will add a longer documentation into a separate PR (page: https://ros-industrial.github.io/ros2_canopen/system-interface.html) in the next day or two.

**Comment by jclinton830** [comment]
Hi! Are there any plans on adding detailed documentation or a user guide on how to use this ROS2 control system interface feature? 

I have checked the official documentation provided but it seems to be empty for now. https://ros-industrial.github.io/ros2_canopen/manual/user-guide/operation/ros2-control-interface.html

**Pull request merged**
